### ISSUE TITLE
Random "Illegal instruction"

### ISSUE THREAD

**Issue by bpiec** [ncw/rclone#1963]
rclone used to crash with `Illegal instruction` message randomly. Now it fails constantly. It happened when I put four big (>3 GB) files that need to be synced.

> What is your rclone version (eg output from `rclone -V`)

rclone v1.39-007-g0d041602β
- os/arch: linux/arm
- go version: go1.9.2

> Which OS you are using and how many bits (eg Windows 7, 64 bit)

armv5tejl armv5tejl armv5tejl GNU/Linux

> Which cloud storage system are you using? (eg Google Drive)

OneDrive

> The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`rclone sync ~/something/somethingelse onedrive:dir --stats 1m0s`

> A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
2018/01/07 00:00:53 DEBUG : Using config file from "/home/username/.rclone.conf"
2018/01/07 00:00:53 DEBUG : rclone: Version "v1.39-007-g0d041602β" starting with parameters ["rclone" "sync" "-vv" "/home/username/something/somethingelse" "onedrive:dir" "--stats" "1m0s"]
2018/01/07 00:00:55 INFO  : One drive root 'dir': Modify window is 1s
2018/01/07 00:00:56 DEBUG : some_file: Size and modification time the same (differ by 0s, within tolerance 1s)
[more messages like this]
Illegal instruction
```

I am pasting output of `cat /proc/cpuinfo` also:

```
Processor       : ARM926EJ-S rev 1 (v5l)
BogoMIPS        : 992.87
Features        : swp half thumb fastmult edsp
CPU implementer : 0x56
CPU architecture: 5TE
CPU variant     : 0x2
CPU part        : 0x131
CPU revision    : 1
Cache type      : write-back
Cache clean     : cp15 c7 ops
Cache lockdown  : format C
Cache format    : Harvard
I size          : 16384
I assoc         : 4
I line length   : 32
I sets          : 128
D size          : 16384
D assoc         : 4
D line length   : 32
D sets          : 128

Hardware        : Feroceon-KW
Revision        : 0000
Serial          : 0000000000000000
```

**Comment by DrPeterVC** [comment]
I too am having similar problems using the same configuration - OneDrive as remote and on ARM32 NAS.  For me it does seem random and associated with file size.

Here is my cpuinfo:
~ # cat /proc/cpuinfo
Processor               : ARM926EJ-S rev 5 (v5l)
BogoMIPS              : 183.09
Features                 : swp half thumb fastmult edsp java
CPU implementer : 0x41
CPU architecture. : 5TEJ
CPU variant           : 0x0
CPU part                : 0x926
CPU revision.        : 5
Cache type            : write-back
Cache clean          : cp15 c7 ops
Cache lockdown  : format C
Cache format       : Harvard
I size                      : 32768
I assoc                  : 4
I line length.         : 32
I sets                     : 256
D size                    : 32768
D assoc                : 4
D line length        : 32
D sets                   : 256

Hardware            : Oxsemi NAS
Revision              : 0000
Serial                    : 00000b6514b6ee80
~ #

**Comment by ncw** [comment]
@DrPeterVC  Thanks for your core dump.  I managed to get the faulting address out of it with gdb, which very rapidly lead to this issue: https://github.com/golang/go/issues/18694

I'm not sure if that issue is fixed or not, but it seems likely that compiling with go-tip might help...

Here is a version of rclone compiled with go-tip can you both give that a go?

[rclone-v1.39-015-g28480c05-arm-go-tip.zip](https://github.com/ncw/rclone/files/1610284/rclone-v1.39-015-g28480c05-arm-go-tip.zip)

**Comment by bpiec** [comment]
This is what I got:

`line 1:  6465 Illegal instruction`

**Comment by bpiec** [comment]
This is funny. I am using script to run sync. The script is very simple:

```
-bash-3.2$ cat ~/scripts/myscript
rclone sync ~/something/somethingelse onedrive:dir --stats 1m0s
```

And when I run it from script, I get the line:column number. But when I run it directly I get dry message only:

```
-bash-3.2$ rclone sync ~/something/somethingelse onedrive:dir --stats 1m0s
Illegal instruction
-bash-3.2$ ~/scripts/myscript
/home/username/scripts/myscript: line 1: 25896 Illegal instruction     rclone sync ~/something/somethingelse onedrive:dir --stats 1m0s
```

Note that the column number is different this time.

**Comment by DrPeterVC** [comment]
Nick - I managed to get the Dev version to fail with illegal instruction too (I have emailed you a new core dump).  It is interesting to note that the go issue you mentioned is on identical hardware MyBookWorld white light as I am using!

**Comment by DrPeterVC** [comment]
I too have discovered something weird about this bug.  If I run rclone and just leave it to run through it seems to work fine but if I run an identical command but just touch my tablet screen (android using JuiceSSH) it comes up with Illegal instruction (see below).  I should point out that the Cron job previously did fail (though it is running ok currently).

Here is the transcript (I renamed rclone to make sure I was running the right version!):
~ # ./rclonedebug copy --dry-run OnedriveRemote:/"My Photos" /shares/Public/"Shared Pictures" --ignore-size
~ # ./rclonedebug copy --dry-run OnedriveRemote:/"My Photos" /shares/Public/"Shared Pictures" --ignore-size
Illegal instruction
~ # ./rclonedebug -V
rclone v1.39-DEV
- os/arch: linux/arm
- go version: devel +a62071a209 Sat Jan 6 04:52:00 2018 +0000
~ #

**Comment by bpiec** [comment]
DrPeterVC: how do you make this core dump?

**Comment by DrPeterVC** [comment]
Simple, just type the following command before you run rclone:

~ # ulimit -c unlimited

The default memory for core dumps is zero (otherwise you would soon run out of space with a faulty program :-) )

**Comment by ncw** [comment]
This is looking like it is a go bug, or something weird about your hardware!

The version I posted (rclone-v1.39-015-g28480c05-arm-go-tip.zip) above should have debugging symbols so gdb should be a bit more useful.  If either of you could attach a core dump made with that - that would be useful.

The other thing to consider is whether either of you could join in golang/go#18694 - namely respond to the go developers requests for debugging?

It looks like you'd have to install gdb on your NAS - not sure how hard that is...

**Comment by DrPeterVC** [comment]
I did send you a core dump using that version.  But I think you are right that this is a strange system configuration issue.  I find that I can consistently make rclone produce an illegal instruction by just tapping the JuiceSSH screen after typing the command!

**Comment by DrPeterVC** [comment]
(cont....). However it previously did fail both without touching the screen and (once only) running under Cron.

As I mentioned before the hardware in issue 18694 is identical to mine (WD MyBookWorld white light) and this NAS does have a reputation for unexplained crashes ( currently mine has been running for over 8 days).

Here is a script from my last touch screen illegal instruction.

~ # ./rclonedebug -V
rclone v1.39-DEV
- os/arch: linux/arm
- go version: devel +a62071a209 Sat Jan 6 04:52:00 2018 +0000
~ # ulimit -c unlimited
~ # ./rclonedebug ls OnedriveRemote:
Illegal instruction (core dumped)
~ #

Please let me know if you want the core dump, if you think it might help and I will zip it up for you.

Reading the golang issue you refer to it does not look like it was actually resolved and was automatically closed a few months ago.

I will investigate how I might get gdb on my NAS.

Thanks for all your help and guidance.

**Comment by ncw** [comment]
> Please let me know if you want the core dump, if you think it might help and I will zip it up for you.

Yes please.  I'll see if gdb works a bit better with it since it will have symbols...

> I will investigate how I might get gdb on my NAS.

I think that would be the most productive thing if you can join in that thread.  I note that go 1.10 is in beta so now would be a great time :-) The release is scheduled for 1st Feb...

**Comment by ncw** [comment]
> I did send you a core dump using that version.

Sorry missed that bit.  The original one you sent me didn't appear to have symbols?  If you can package up the rclone binary and the core dump in a zip that would be great.  You can attach it here if you want or email it to me since it likely contains sensitive stuff.

**Comment by DrPeterVC** [comment]
Ok sorry for the delay I have been away.

I did manage to get a version of gdb (via optware) on my NAS but cannot make any sense of the output - any ideas???
~ # gdb ./rclonedebug core.8848


dlopen failed on 'libthread_db.so.1' - libthread_db.so.1: cannot open shared object file: No such file or directory
GDB will not be able to debug pthreads.

GNU gdb 6.8
Copyright (C) 2008 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "arm-none-linux-gnueabi"...
Dwarf Error: wrong version in compilation unit header (is 4, should be 2) [in module /root/rclonedebug]

warning: core file may not match specified executable file.
Core was generated by `rclone -V'.
Program terminated with signal 4, Illegal instruction.
[New process 8848]
[New process 8851]
[New process 8850]
[New process 8849]
#0  0xffff0514 in ?? ()
(gdb)

**Comment by DrPeterVC** [comment]
Hot news!  I have just installed go on my pc then cross compiled the hello world program and got it to fail with illegal instruction in exactly the same way (touching the tablet screen).  So my next task is to dig out my raspberry pi and see if it is a problem there too ...  Watch this space.

**Comment by DrPeterVC** [comment]
OK so that confirms it.  It is a problem with my NAS (using go at least).
Tried the following simple program:

package main

import (
        "fmt"
        "time"
)

func main() {
    time.Sleep(10*time.Second)
    fmt.Printf("hello, world\n")
}

Built it for Linux-ARM32(5).  It ran fine on both the NAS and the Pi.  However if I touched the screen within the 10 seconds (this brings up the special characters (HOME, END arrows etc...) in JuiceSSH when running on the pi there was no error but on the NAS it caused an illegal instruction termination. 

Many of these home consumer NAS boxes are built down to a price and their firmware is notoriously flakey.  The WD MyBookWorld is known for its random lock ups.  I recently had to replace the hard drive - which turns out to be one that WD themselves do not recommend for use in servers!!!!  

I had originally planned to build my own replacement NAS with my Pi and a separate drive - but decided against as the Pi has no SATA port and is limited to USB2 - instead I decided to buy a second hand hard replacement drive (£25 - ebay) and resurrected my WD - NAS, but now I only use it to back up data which exists elsewhere (hence why I was so keen to use rclone).

Anyway if you have read this far - thanks for your time - and apologies to ncw (Nick) for wasting time on looking for potential problems in his excellent utility.

I do not plan to post anything further in this thread.

**Comment by ncw** [comment]
@DrPeterVC wrote:
> OK so that confirms it. It is a problem with my NAS (using go at least).

Thanks for doing the investigation and narrowing down the problem.  It might be worth reporting your investigations on https://github.com/golang/go/issues/18694

> The WD MyBookWorld is known for its random lock ups.

My experience with flakey consumer stuff is that it isn't cooled properly - I don't know if you can add an extra fan, or run with the case off...

**Comment by ncw** [comment]
The fix for the above issue got into go 1.10 and hence into rclone 1.40.  I'm going to close this now - please re-open if you still have problems.
